src/infill.cpp

```diff
@@ -48,12 +52,127 @@ void Infill::generate(Polygons& result_polygons, Polygons& result_lines)
     case EFillMethod::ZIG_ZAG:
         generateZigZagInfill(result_lines, line_distance, fill_angle, connected_zigzags, use_endpieces);
         break;
+    case EFillMethod::TRUNCATED_OCTAHEDRON:
+        generateTroctInfill(result_polygons, fill_angle);
+        generateTroctInfill(result_polygons, fill_angle + 90);
+        break;
     default:
         logError("Fill pattern has unknown value.\n");
         break;
     }
 }
 
+void Infill::generateTroctInfill(Polygons& result, double rotation)
+  {
+    int extrusionWidth = infill_line_width;
+    int lineSpacing = line_distance;
+    int infillOverlap = infill_overlap;
+    int posZ = z;
+
+    Polygons outline = in_outline.offset(extrusionWidth * infillOverlap / 100);
+    PointMatrix matrix(rotation);
+    outline.applyMatrix(matrix);
+    AABB boundary(outline);
+
+    // ignore infill for areas smaller than line spacing
+    if((abs(boundary.min.X - boundary.max.X) + abs(boundary.min.Y - boundary.max.Y)) < lineSpacing){
+      return;
+    }
+
+    // fix to normalise against diagonal infill
+    lineSpacing = lineSpacing * 2;
+
+    uint64_t Zscale = SQRT2MUL(lineSpacing);
+
+    int offset = abs(posZ % (Zscale) - (Zscale/2)) - (Zscale/4);
+    boundary.min.X = ((boundary.min.X / lineSpacing) - 1) * lineSpacing;
+    boundary.min.Y = ((boundary.min.Y / lineSpacing) - 1) * lineSpacing;
+
+    unsigned int lineCountX = (boundary.max.X - boundary.min.X + (lineSpacing - 1)) / lineSpacing;
+    unsigned int lineCountY = (boundary.max.Y - boundary.min.Y + (lineSpacing - 1)) / lineSpacing;
+    int rtMod = int(rotation / 90) % 2;
+    // with an odd number of lines, sides need to be swapped around
+    if(rtMod == 1){
+      rtMod = (lineCountX + int(rotation / 90)) % 2;
+    }
+
+    // draw non-horizontal walls of octohedrons
+    Polygons po;
+    PolygonRef p = po.newPoly();
+    for(unsigned int ly=0; ly < lineCountY;){
+      for(size_t it = 0; it < 2; ly++, it++){
+        int side = (2*((ly + it + rtMod) % 2) - 1);
+        int y = (ly * lineSpacing) + boundary.min.Y + lineSpacing / 2 - (offset/2 * side);
+        int x = boundary.min.X-(offset/2);
+        if(it == 1){
+          x = (lineCountX * (lineSpacing)) + boundary.min.X + lineSpacing / 2 - (offset/2);
+        }
+        p.add(Point(x,y));
+        for(unsigned int lx=0; lx < lineCountX; lx++){
```

Comment:
This loop seems to do its very best to prevent some minor code duplication but as an end result there's so much magic going on that simply listing the 8 sides of the polygon in order would be much simpler to read and easier to maintain.